### ISSUE TITLE
Calculating bam read depth

### ISSUE THREAD

**Issue by MrKevinDC** [samtools/htsjdk#1096]
Hi for the first time, I am trying to find a way to use Samtools' depth function in a Javascript Genome Browser I have been tasked with working on in my Bioinformatics MSc (first time programming). So I am having a hard time finding in these Java API packages that function or anything similar to it. Has it been implemented in some way, or should I try something else? I am quite new to programming and probably this is not what is considered an issue but I don't know where to go to...
Thanks for any responses!

**Comment by yfarjoun** [comment]
perhaps you want to use SamLocusIterator? that will give you a locusInfo
which will include all the reads that overlap that locus and you'll be able
to get the depth from there.

For a use-case you can take a look at picard's CollectWgsMetrics, for
example.

That said a "PileupDepth" example in htsjdk might be a good idea...

**Comment by MrKevinDC** [comment]
Thank you very much @yfarjoun! I've seen you answering other threads and I have to say you are providing very useful help. I'll check out those options

**Comment by lbergelson** [comment]
As a heads up, htsjdk is a *java* library, not a *javascript* library.  So you won't be able to easily use htsjdk in a javascript browser app.

**Comment by MrKevinDC** [comment]
Yeah I read about the Nashorn engine and I was planning on importing Java classes from the Javascript, hopefully it is the way to go

**Comment by MrKevinDC** [comment]
Java classes *to the Javascript

**Comment by MrKevinDC** [comment]
I´m open to any suggestions

**Comment by lindenb** [comment]
you cannot use nahsorn inside a browser like firefox or chrome because their javascript engines are not java-based. You'd better have a look at some "pure" javascript bam/sam parser like: https://github.com/dasmoth/dalliance ...

**Comment by MrKevinDC** [comment]
Oh, thank you @lindenb  there are many things that still overwhelm with this task. I´ll check it out. Thanks for everything guys!

**Comment by tfenne** [comment]
Closing as stale — sorry this never got a proper answer.

The htsjdk equivalent of `samtools depth` is `htsjdk.samtools.util.SamLocusIterator`, which walks a SAM/BAM locus by locus and hands you the reads piled up at each position. Note it's a Java library, so it won't be directly usable from a JavaScript genome browser without a server-side component.
